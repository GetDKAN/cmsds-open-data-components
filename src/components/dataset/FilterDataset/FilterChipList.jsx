import React from 'react'
import FilterChip from './FilterChip'
import FontAwesomePro from '../../common/FontAwesomePro/FontAwesomePro'
import PropTypes from 'prop-types'
import './FilterChip.scss'

const FilterChipList = ({
  filters,
  deleteFilter,
  resetFilters,
  hidden,
  resetHiddenColumns,
  reordered,
  resetColumnOrder
}) => {
  const showReset = filters.length > 0 || hidden || reordered
  const applyReset = () => {
    resetHiddenColumns()
    resetColumnOrder()
    resetFilters()
  }
  // lets avoid displaying filters if the only one there is our empty starter filter
  return (
    <div className='chip-list'>
      {filters && filters.map((filter, i) => (
        <FilterChip key={`${i}-${filter.div}`} filter={filter} deleteFilter={deleteFilter} i={i} />
      ))}
      {hidden && (
        <button className='chip-body' onClick={resetHiddenColumns}>
          <FontAwesomePro icon='columns' size={12} fill='#323A45' aria-hidden={true} />
          Columns hidden
          <FontAwesomePro icon='times' size={12} fill='#6F757C' aria-hidden={true} />
        </button>
      )}
      {reordered && (
        <button className='chip-body' onClick={resetColumnOrder}>
          <FontAwesomePro icon='columns' size={12} fill='#323A45' aria-hidden={true} />
          Columns reordered
          <FontAwesomePro icon='times' size={12} fill='#6F757C' aria-hidden={true} />
        </button>
      )}
      {showReset && (
        <button className='clear-all' onClick={applyReset}>Clear all filters</button>
      )}
    </div>
  )
}

FilterChipList.propTypes = {
  /**
   * Array of filters
   */
  filters: PropTypes.arrayOf(PropTypes.shape({
    column: PropTypes.string,
    condition: PropTypes.oneOf([
      '=',
      '<>',
      '>',
      '<',
      'LIKE',
      'starts with',
      'is_empty',
      'not_empty',
      '' // We need to allow an empty case because there are initially no filters so there is no condition set which throws a type error
    ]),
    value: PropTypes.string
  })),
  /**
   * `true` if some columns have been hidden
   * as a result of a filter
   */
  hidden: PropTypes.bool,
  /**
   * `true` if columns have been reordered
   */
  reordered: PropTypes.bool,
  /**
   * Callback function to delete filter
   */
  deleteFilter: PropTypes.func,
  /**
   * Callback function to reset the original column order
   */
  resetColumnOrder: PropTypes.func,
  /**
   * Callback function to reset hidden columns
   */
  resetHiddenColumns: PropTypes.func,
  /**
   * Callback function to remove all applied filters
   */
  resetFilters: PropTypes.func
}
FilterChipList.displayName = 'FilterChipList'
export default FilterChipList
