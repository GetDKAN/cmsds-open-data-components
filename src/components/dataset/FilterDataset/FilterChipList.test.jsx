import { render, fireEvent, screen } from '@testing-library/react'
import renderer from 'react-test-renderer'
import FilterChipList from './FilterChipList'

const componentArgs = {
  filters: [
    {
      column: 'col4',
      condition: 'is_empty',
      value: ''
    },
    {
      column: 'col2',
      condition: 'LIKE',
      value: '%medicare%'
    }
  ],
  hidden: false,
  reordered: false,
  deleteFilter: vi.fn(),
  resetColumnOrder: vi.fn(),
  resetHiddenColumns: vi.fn(),
  resetFilters: vi.fn()
}

describe('FilterChipList component.', () => {
  it('Matches snapshot.', () => {
    const renderedFilterChipList = renderer.create(
      <FilterChipList {...componentArgs} />
    ).toJSON()

    expect(renderedFilterChipList).toMatchSnapshot()
  })

  it('Renders the \'Columns hidden\' chip when \'hidden\' prop is set.', () => {
    const resetHiddenColumns = vi.fn()
    
    render(
      <FilterChipList
        {...componentArgs}
        hidden={true}
        resetHiddenColumns={resetHiddenColumns}
      />
    )

    expect(screen.getByRole('button', { name: 'Clear all filters' })).toBeInTheDocument()
    expect(screen.getByRole('button', { name: 'Columns hidden' })).toBeInTheDocument()

    fireEvent.click(screen.getByRole('button', { name: 'Columns hidden' }))

    expect(resetHiddenColumns).toHaveBeenCalled()
  })

  it('Renders the \'Columns reordered\' chip when \'reordered\' prop is set.', () => {
    const resetColumnOrder = vi.fn()
    
    render(
      <FilterChipList
        {...componentArgs}
        reordered={true}
        resetColumnOrder={resetColumnOrder}
      />
    )

    expect(screen.getByRole('button', { name: 'Columns reordered' })).toBeInTheDocument()

    fireEvent.click(screen.getByRole('button', { name: 'Columns reordered' }))

    expect(resetColumnOrder).toHaveBeenCalled()
  })

  it('All filters are removed when \'Clear all filters\' is clicked.', () => {
    const resetColumnOrder = vi.fn()
    const resetHiddenColumns = vi.fn()
    const resetFilters = vi.fn()
    
    render(
      <FilterChipList
        {...componentArgs}
        resetColumnOrder={resetColumnOrder}
        resetHiddenColumns={resetHiddenColumns}
        resetFilters={resetFilters}
      />
    )

    expect(screen.getByRole('button', { name: 'Clear all filters' })).toBeInTheDocument()

    fireEvent.click(screen.getByRole('button', { name: 'Clear all filters' }))

    expect(resetColumnOrder).toHaveBeenCalled()
    expect(resetHiddenColumns).toHaveBeenCalled()
    expect(resetFilters).toHaveBeenCalled()
  })

  it('Renders the \'Clear all filters\' button when there are no filters but hidden columns are set.', () => {
    render(
      <FilterChipList
        {...componentArgs}
        filters={[]}
        hidden={true}
      />
    )

    expect(screen.getByRole('button', { name: 'Clear all filters' })).toBeInTheDocument()
  })

  it('Renders the \'Clear all filters\' button when there are no filters but reordered columns are set.', () => {    
    render(
      <FilterChipList
        {...componentArgs}
        filters={[]}
        reordered={true}
      />
    )

    expect(screen.getByRole('button', { name: 'Clear all filters' })).toBeInTheDocument()
  })
})
