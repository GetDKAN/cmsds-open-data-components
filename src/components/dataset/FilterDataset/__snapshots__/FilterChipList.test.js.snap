// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterChipList component. Matches snapshot. 1`] = `
<div
  className="chip-list"
>
  <button
    aria-label="Remove col4 is empty filter"
    className="chip-body"
    onClick={[Function]}
  >
    <svg
      aria-hidden={true}
      aria-label="filter"
      className="fa fa-filter"
      focusable="false"
      height={12}
      role="img"
      viewBox="0 0 512 512"
      width={12}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M463.952 0H48.057C5.419 0-16.094 51.731 14.116 81.941L176 243.882V416c0 15.108 7.113 29.335 19.2 40l64 47.066c31.273 21.855 76.8 1.538 76.8-38.4V243.882L497.893 81.941C528.042 51.792 506.675 0 463.952 0zM288 224v240l-64-48V224L48 48h416L288 224z"
      />
    </svg>
    "
    col4
    " Is empty 
    <svg
      aria-hidden={true}
      aria-label="times"
      className="fa fa-times"
      focusable="false"
      height={12}
      role="img"
      viewBox="0 0 384 512"
      width={12}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M231.6 256l130.1-130.1c4.7-4.7 4.7-12.3 0-17l-22.6-22.6c-4.7-4.7-12.3-4.7-17 0L192 216.4 61.9 86.3c-4.7-4.7-12.3-4.7-17 0l-22.6 22.6c-4.7 4.7-4.7 12.3 0 17L152.4 256 22.3 386.1c-4.7 4.7-4.7 12.3 0 17l22.6 22.6c4.7 4.7 12.3 4.7 17 0L192 295.6l130.1 130.1c4.7 4.7 12.3 4.7 17 0l22.6-22.6c4.7-4.7 4.7-12.3 0-17L231.6 256z"
      />
    </svg>
  </button>
  <button
    aria-label="Remove col2 contains medicare filter"
    className="chip-body"
    onClick={[Function]}
  >
    <svg
      aria-hidden={true}
      aria-label="filter"
      className="fa fa-filter"
      focusable="false"
      height={12}
      role="img"
      viewBox="0 0 512 512"
      width={12}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M463.952 0H48.057C5.419 0-16.094 51.731 14.116 81.941L176 243.882V416c0 15.108 7.113 29.335 19.2 40l64 47.066c31.273 21.855 76.8 1.538 76.8-38.4V243.882L497.893 81.941C528.042 51.792 506.675 0 463.952 0zM288 224v240l-64-48V224L48 48h416L288 224z"
      />
    </svg>
    "
    col2
    " Ｃ 
    medicare
    <svg
      aria-hidden={true}
      aria-label="times"
      className="fa fa-times"
      focusable="false"
      height={12}
      role="img"
      viewBox="0 0 384 512"
      width={12}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M231.6 256l130.1-130.1c4.7-4.7 4.7-12.3 0-17l-22.6-22.6c-4.7-4.7-12.3-4.7-17 0L192 216.4 61.9 86.3c-4.7-4.7-12.3-4.7-17 0l-22.6 22.6c-4.7 4.7-4.7 12.3 0 17L152.4 256 22.3 386.1c-4.7 4.7-4.7 12.3 0 17l22.6 22.6c4.7 4.7 12.3 4.7 17 0L192 295.6l130.1 130.1c4.7 4.7 12.3 4.7 17 0l22.6-22.6c4.7-4.7 4.7-12.3 0-17L231.6 256z"
      />
    </svg>
  </button>
  <button
    className="clear-all"
    onClick={[Function]}
  >
    Clear all filters
  </button>
</div>
`;
